src/main/scala/datalog/execution/ir/IRTreeGenerator.scala

```diff
@@ -98,43 +97,124 @@ class IRTreeGenerator(using val ctx: InterpreterContext)(using JITOptions) {
                     ScanOp(r, DB.Derived, KNOWLEDGE.Known)
                 }): _*
               )
-            }):_*
+            }): _*
           )
       case _ =>
         debug("AST node passed to semiNaiveEval:", () => ctx.storageManager.printer.printAST(ast))
         throw new Exception("Wrong ASTNode received when generating naive IR")
     }
   }
 
-  def generateNaive(ast: ASTNode): IROp[Any] = {
+  /**
+   * A step in the strata generation. At each step, some new relations should
+   * be derived, and some relations should be kept from the previous strata.
+   *
+   * @param toDerive The relations to derive in this step.
+   * @param toKeep   The relations to keep from the previous strata.
+   * @param rules    The rules to derive in this step.
+   */
+  private case class Step(toDerive: Set[RelationId],
+                          toKeep: Set[RelationId],
+                          rules: mutable.Map[Int, ASTNode])
+
+  /**
+   * Stratifies the program into a sequence of steps, where strata are rules
+   * that should be derived simultaneously.
+   *
+   * @param ast The root of the program.
+   * @return A sequence of steps for which relations should be derived.
+   */
+  private def stratify(ast: ASTNode): Seq[Step] = {
     ast match {
       case ProgramNode(ruleMap) =>
-        DoWhileOp(
-          DB.Derived,
-          SequenceOp(OpCode.LOOP_BODY,
-            SwapAndClearOp(),
-            naiveEval(ruleMap)
-          )
-        )
+        val components = ctx.precedenceGraph.scc(ctx.toSolve)
+        val kept = components.scanLeft(Set[Int]())(_ ++ _)
+        components.zip(kept).map((ids, keep) =>
+          val rules = ruleMap.filter((r, _) => ids.contains(r))
+          Step(ids, keep, rules))
       case _ => throw new Exception("Non-root passed to IR Program")
     }
   }
 
+  /**
+   * Generates the IR for the naive evaluation strategy.
+   *
+   * @param ast The root of the program.
+   * @return The IR for the naive evaluation strategy.
+   */
+  def generateNaive(ast: ASTNode): IROp[Any] = {
+    val steps = stratify(ast).map(step =>
+      val keep =
+        SequenceOp(OpCode.OTHER,
+          step.toKeep.toSeq.map(r =>
+            InsertOp(r, DB.Derived, KNOWLEDGE.New,
+              ScanOp(r, DB.Derived, KNOWLEDGE.Known)
+                .asInstanceOf[IROp[Any]])
+          ): _*
+        )
+      val insertions = SequenceOp(
+        OpCode.EVAL_NAIVE,
+        step.rules.map((rId, rule) =>
+          val res = naiveEvalRule(rule).asInstanceOf[IROp[Any]]
+          InsertOp(rId, DB.Derived, KNOWLEDGE.New, res)
+        ).toSeq: _*
+      )
+      DoWhileOp(
+        DB.Derived,
+        SequenceOp(OpCode.LOOP_BODY,
+          SwapAndClearOp(),
+          keep,
+          insertions
+        )
+      )
+    )
+    SequenceOp(
```

Comment:
You're missing a `ProgramOp` at the topmost level. You can modify it so that it takes in a list of operations instead of a single one (just make sure you update the pretty printer to reflect that).

I would expect the structure of the generated code to look something like:

```
ProgramOp [
    SequenceOp { OpCode.STRATA, [
        DoWhileOp(
            DB.Derived,
            SequenceOp(OpCode.LOOP_BODY,
                SwapAndClearOp(),
                naiveEval(ruleMap <*of the rules for the correct strata*>) // this should generate almost identical code to the non-stratified version, but only involve the relations in the first strata
            )
        ),
        SwapAndClearOp() // might need a more specific operation if there are relations that you have derived in the previous strata that you *do not want* as input to the next strata, but I'm not sure if that ever happens. The point is to copy "New" IDBs into "Derived" IDBs so we can treat the result of the previous strata as input to the next one
    ),
    SequenceOp { OpCode.STRATA, [
        DoWhileOp( ...
```